Python cffi

mcoram opened the pull request:
So, I've been otherwise engaged.... Happy new year.
This pull request represents where I'm at for python cffi and opens discussion. Sorry if my github usage is sloppy.
C.f. https://github.com/alexameen/cpoke/issues/6

Basically I've got a semi-automatic process (in need of refinement) that pulls together the headers, specifically enum/struct definitions and function prototypes into a cffi/include/all.E file.  This file is good enough that cffi imports it but there are noteworthy hacks in place. E.g. I've manually deleted various too-complicated parts, inline functions, many static data declarations, etc. I did a little macro expansion by hand, though I've tried using the preprocessor too. On the pull request you could run 'make cffi/data/all.E' and compare that all.E with the cffi/include/all.E to see the kind of futzing I did in more detail.

Anyway, to warm up things, run this:
  make cffi/cpoke.so

Then, the demo that's available, such as it is, is just:
  cd cffi && python example2.py

As you'll see, I haven't gotten too far getting things to run, though I haven't had much time yet with the updated all.E. Basically I have working "tests" of:
  lib.get_cp_from_stats
  lib.cstore_get_pokemon
Where the latter uses a trivial helper function to return &CSTORE that I hacked into test_cstore.c for now. Maybe a cffi.c somewhere would be more appropriate for these bridge functions.

If you run 'dir(lib)' on the constructed object in the python interpreter that comes up, you can see the subset of c-objects that have wrappers. There's quite a bit there, but I don't know how much of it is really workable. E.g. I've gotten some segfaults trying to do lib.cstore_get_pokemon_by_name(cstore_p, name, mon_pp).

Commit:
experimental cffi with one working test call.

mcoram committed:
Get cstore to import (mostly).

mcoram committed:
semi-automatic all.E creation and example2.py

mcoram committed:
macro expansion fiddling and static data definition, though it looks useless.

alexameen commented:
This looks awesome! I'm going to experiment with it a bit and merge soon.

Thank you for your help

mcoram commented:
Cool. I got a bit more to work but there’s still some seg faults I can’t
figure out (e.g. cstore_get_pokemon_by_name)

On Mon, Jan 4, 2021 at 6:38 PM Alex Ameen <notifications@github.com> wrote:

> This looks awesome! I'm going to experiment with it a bit and merge soon.
>
> Thank you for your help
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexameen/cpoke/pull/45#issuecomment-754348641>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQYPWYCHIWETD7EWULS5DSYJ3Q3ANCNFSM4VSIUP5Q>
> .
>

alexameen commented:
Hey I know things aren't finished, so you might be aware of this issue; but when trying to build:
```
make cffi/data/all.E;
make cffi/cpoke.so;
```
I get a ton of errors linking : `ld: FILENAME.o: relocation R_X86_64_32 against \`.rodata' can not be used when making a shared object; recompile with -fPIC`.

Were you getting that on your builds?

Did I build incorrectly?

mcoram commented:
Interesting. I just did make clean && make cffi/cpoke.so and I didn't see
errors. There are a number of things to track down here.  [e.g. if I'm
sync'd, or if my gcc invokation is using the BSD ld behind the scenes or
what, etc.]
Thanks for mentioning it. The -fPIC switch seems reasonable, but I wonder
why I haven't seem to have needed it or if I'm merely having silent
problems that are undetected. Happy for suggestions.

On Wed, Jan 6, 2021 at 5:53 PM Alex Ameen <notifications@github.com> wrote:

> Hey I know things aren't finished, so you might be aware of this issue;
> but when trying to build:
>
> make cffi/data/all.E;
> make cffi/cpoke.so;
>
> I get a ton of errors linking : ld: FILENAME.o: relocation R_X86_64_32
> against '.rodata' can not be used when making a shared object; recompile
> with -fPIC.
>
> Were you getting that on your builds?
>
> Did I build incorrectly?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexameen/cpoke/pull/45#issuecomment-755832199>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQYPTVW5EVQDMO2YZRII3SYUHZHANCNFSM4VSIUP5Q>
> .
>

alexameen commented:
My bet is that your version/build of `make` is implicitly adding `-fPIC` to `CFLAGS` when you attempt to build a Shared Object.
Another explanation might be that your build of GCC or environment variables add `-fPIC` to all builds.

alexameen commented:
Any idea how to get `cffi_exp_cstore_inc.h`?

alex-ameen committed:
Makefile can generate ORDERED_E automatically, added -fPIC options, expanded clean command

alex-ameen committed:
Update Makefile

mcoram commented:
I don't think it's needed. The first example.py is basically deprecated but
I haven't gotten around to cleanup. I think the cffi/include/all.E (which
makes all.E2, sorry for the hacks) and which resembles but isn't identical
to the cffi/data/all.E, is all that's needed now.

On Wed, Jan 6, 2021 at 7:00 PM Alex Ameen <notifications@github.com> wrote:

> Any idea how to get cffi_exp_cstore_inc.h?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexameen/cpoke/pull/45#issuecomment-755851372>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQYPSUT46MDWJUYZQDLU3SYUPULANCNFSM4VSIUP5Q>
> .
>

mcoram committed:
Merge pull request #1 from alexameen/python-cffi

Makefile can generate ORDERED_E automatically, added -fPIC options

mcoram committed:
Cleanup removing example.py and adding misc experimental hacks.

alexameen commented:
Hey I pushed code to load AI from Shared Objects, and added 2 new AI interface functions.

In practice trying to handle "forced shield/swap" scenarios with the one function was going to be complicated for users. I added two prototypes for extra decision functions, but haven't added them to the battle handler yet.

I wanted to push though so you could see how the AI Shared Object loader worked, and wanted to get those new functions added so you wouldn't need to redo things later.

mcoram commented:
Thanks. One thing I wanted to ask. I'm trying to look for a cause of some
segfaults. You know your macro DEFINE_ENUM_WITH_FLAGS, e.g.
```
DEFINE_ENUM_WITH_FLAGS( store_flag,
    SF_NONE, SF_WRITABLE, SF_THREAD_SAFE, SF_OFFICIAL_DATA, SF_CUSTOM_DATA,
    SF_EXPORTABLE, SF_STANDARD_KEY, SF_TYPED, SF_GET_STRING,
    SF_GET_TYPED_STRING
  );
```
That expands to:
```
typedef enum __attribute__((packed)) { SF_NONE , SF_WRITABLE ,
SF_THREAD_SAFE , SF_OFFICIAL_DATA , SF_CUSTOM_DATA , SF_EXPORTABLE ,
SF_STANDARD_KEY , SF_TYPED , SF_GET_STRING , SF_GET_TYPED_STRING }
store_flag_t; typedef enum __attribute__((packed)) { SF_NONE_M = ( ( ( !! (
SF_NONE ) ) << ( SF_NONE ) ) >> 1 ) , SF_WRITABLE_M = ( ( ( !! (
SF_WRITABLE ) ) << ( SF_WRITABLE ) ) >> 1 ) , SF_THREAD_SAFE_M = ( ( ( !! (
SF_THREAD_SAFE ) ) << ( SF_THREAD_SAFE ) ) >> 1 ) , SF_OFFICIAL_DATA_M = (
( ( !! ( SF_OFFICIAL_DATA ) ) << ( SF_OFFICIAL_DATA ) ) >> 1 ) ,
SF_CUSTOM_DATA_M = ( ( ( !! ( SF_CUSTOM_DATA ) ) << ( SF_CUSTOM_DATA ) ) >>
1 ) , SF_EXPORTABLE_M = ( ( ( !! ( SF_EXPORTABLE ) ) << ( SF_EXPORTABLE ) )
>> 1 ) , SF_STANDARD_KEY_M = ( ( ( !! ( SF_STANDARD_KEY ) ) << (
SF_STANDARD_KEY ) ) >> 1 ) , SF_TYPED_M = ( ( ( !! ( SF_TYPED ) ) << (
SF_TYPED ) ) >> 1 ) , SF_GET_STRING_M = ( ( ( !! ( SF_GET_STRING ) ) << (
SF_GET_STRING ) ) >> 1 ) , SF_GET_TYPED_STRING_M = ( ( ( !! (
SF_GET_TYPED_STRING ) ) << ( SF_GET_TYPED_STRING ) ) >> 1 ) }
store_flag_mask_t; typedef union { store_flag_mask_t mask; struct
__attribute__((packed)) { unsigned int SF_WRITABLE : 1 ; unsigned int
SF_THREAD_SAFE : 1 ; unsigned int SF_OFFICIAL_DATA : 1 ; unsigned int
SF_CUSTOM_DATA : 1 ; unsigned int SF_EXPORTABLE : 1 ; unsigned int
SF_STANDARD_KEY : 1 ; unsigned int SF_TYPED : 1 ; unsigned int
SF_GET_STRING : 1 ; unsigned int SF_GET_TYPED_STRING : 1; }; }
store_flag_flags_t
```
cffi can't consume that because it only seems to accept constants in the
enum definition (not compile-time constants). I've been quietly hacking
that to:
```typedef enum {    SF_NONE, SF_WRITABLE, SF_THREAD_SAFE,
SF_OFFICIAL_DATA, SF_CUSTOM_DATA,
    SF_EXPORTABLE, SF_STANDARD_KEY, SF_TYPED, SF_GET_STRING,
    SF_GET_TYPED_STRING} store_flag_t;
typedef unsigned int store_flag_mask_t;

```
In particular, I make store_flag_mask_t and unsigned int. But I don't
actually know that that's the correct storage type for the enum in the
compiled code. Can you think of any simple fixes? I guess I was thinking of
extending the macro to add a "print" function that I could compile and run
to emit the compiled version of the store_flag_mask_t. Shrug.

On Fri, Jan 8, 2021 at 10:40 PM Alex Ameen <notifications@github.com> wrote:

> Hey I pushed code to load AI from Shared Objects, and added 2 new AI
> interface functions.
>
> In practice trying to handle "forced shield/swap" scenarios with the one
> function was going to be complicated for users. I added two prototypes for
> extra decision functions, but haven't added them to the battle handler yet.
>
> I wanted to push though so you could see how the AI Shared Object loader
> worked, and wanted to get those new functions added so you wouldn't need to
> redo things later.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexameen/cpoke/pull/45#issuecomment-757106073>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQYPQD7WZH7EZHGMT4GJDSY727HANCNFSM4VSIUP5Q>
> .
>

alexameen commented:
Its likely made because i used a packed enum. It creates an unusual size for the data type. Let me write back in a minute with more info

alexameen commented:
If I need to use the expanded forms in my sources let me know it's not an issue. I was just trying to be tricky; the macros are not necessary.

mcoram committed:
Current status.

mcoram committed:
Current status.

mcoram committed:
resolve conflicts.

mcoram committed:
Merge branch 'python-cffi' of https://github.com/mcoram/cpoke into python-cffi
